### PR TITLE
Add scrapy-autoextract to Pypi

### PR DESCRIPTION
**Pull request by croqaz**
This PR:

* adds a few basic tests for the middleware
* use tox to test agains multiple Python versions
* adds Travis CI integration to publish tags on Pypi automatically

### CODE REVIEW

**Comment by croqaz** [comment]
All tests pass, I believe it's ready to merge

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[tox]
+envlist = py36, py37
```
Comment (kmike) [inline tox.ini]: Can we support Python 3.5 as well?

File: README.rst [hunk 2]
```diff
@@ -8,6 +8,16 @@ The middleware adds the result of AutoExtract to ``response.meta['autoextract']`
...
 
 
+Installation
+===========
+
+::
+
+    pip install scrapy-autoextract
+
+scrapy-autoextract requires Python 3.
```
Comment (kmike) [inline README.rst]:
```suggestion
scrapy-autoextract requires Python 3.5+
```

(or 3.6+, depending on what we decide to support)

File: scrapy_autoextract/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from .middlewares import AutoExtractMiddleware  # noqa: F401
```
Comment (kmike) [inline scrapy_autoextract/__init__.py]: 👍

Comment (kmike) [reply scrapy_autoextract/__init__.py]: This should allow to simplify README ("Configuration" section)

File: test/test_autoextract.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import os
+import sys
+import pytest
+
+from w3lib.http import basic_auth_header
+from scrapy.http import Request, Response
+from scrapy.spiders import Spider
+from scrapy.utils.test import get_crawler
+
+sys.path.append(os.getcwd())
```
Comment (kmike) [inline test/test_autoextract.py]: why is this line needed? Tests should run for installed scrapy_autoextract. Also, os.getcwd() looks incorrect, because you may invoke these tests while the current folder is arbitrary; cwd is nott necessarily related to path to this file.

File: test/test_autoextract.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    assert out is None
+    assert req.meta.get('autoextract') is None
+    res = Response(req.url, request=req)
+    assert mw.process_response(req, res, spider) == res
+
+
+def _assert_enabled(spider,
+                    settings=None,
+                    url='http://quotes.toscrape.com',
+                    proxyurl='autoextract.scrapinghub.com',
```
Comment (kmike) [inline test/test_autoextract.py]: probably makes sense to rename "proxy..." arguments to something else, as AutoExtract is not a proxy

File: test/test_autoextract.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                    proxyauth=basic_auth_header('apikey', '')):
```
Comment (kmike) [inline test/test_autoextract.py]: this is not a value of Proxy-Auth header

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+[testenv]
+deps =
+    -r test/requirements.txt
+
+commands =
+    flake8 scrapy_autoextract
```
Comment (kmike) [inline tox.ini]: I think it is better to have flake8 as a separate tox environment

File: tox.ini (hunk 1 above)
Comment (croqaz) [reply tox.ini]: We should look into the future, not the past, now that Python 3.8 was released :)
But yes, I'll add Python 3.5 :)

**Comment by croqaz** [comment]
Ok the build for Python 3.5 is failing because of JSON and str vs bytes.
I'll install Python 3.5 tomorrow and fix it.

**Comment by croqaz** [comment]
Fixed the tests for Python 3.5.
`json.loads(b'[{}]')` doesn't work in 3.5, but the body is required to be bytes, by the Scrapy Response object.

I believe I fixed all comments. Ready to merge 🚀

File: .travis.yml [hunk 8]
```diff
@@ -14,7 +14,9 @@ matrix:
 install: pip install -U tox
 
 # command to run tests
-script: tox
+script:
+  - tox -e flake8
```
Comment (kmike) [inline .travis.yml]: you can put it to matrix, as another env

File: test/test_autoextract.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    mw = _mock_mw(spider, settings)
+
+    req = Request(url, meta=AUTOX_META)
+    out = mw.process_request(req, spider)
+    assert api_url in out.url
+    assert out.meta['autoextract'].get('enabled')
+    assert out.headers.get('Authorization') == api_auth
+
+    resp = Response(out.url, request=out)
+    resp._body = '[{}]'  # workaround to make the body work with Python 3.5+
```
Comment (kmike) [inline test/test_autoextract.py]: It doesn't look good; I think tests showed you a real issue in Python 3.5. In Scrapy response.body is bytes, response.text is text. In case of JSON you probably need to call `response.body.decode('utf8')`, and pass it to json.loads.

File: .travis.yml (hunk 8 above)
Comment (croqaz) [reply .travis.yml]: Ok, I managed to do this, it took some time, but... what's the advantage? That moving this to another env will run the flake8 only once, just like a normal test?

Comment (kmike) [reply .travis.yml]: Yes; also, you won't have all tests failing if that's just a style issue. From job builds it'd be clear if this is a general issue, a problem with flake check, a problem with Python 3.5 only, etc.

**Comment by kmike** [comment]
thanks @croqaz!
